libgnucash/quotes/CMakeLists.txt

```diff
@@ -1,3 +1,54 @@
+### libgnc-quotes
+set (quotes_SOURCES
+    gnc-quotes.cpp
+)
+
+# Add dependency on config.h
+set_source_files_properties (${quotes_SOURCES} PROPERTIES OBJECT_DEPENDS ${CONFIG_H})
+
+set(quotes_noinst_HEADERS
+    gnc-quotes.hpp
+)
+
+add_library(gnc-quotes ${quotes_SOURCES} ${quotes_noinst_HEADERS})
```

Comment:
Good point. My reasoning was to keep all  code for quote handling together.  There are also the perl scripts that go with this.

Some background as to where I'd like this to go:

There are 4 perl scripts:  `gnc-fq-update`, `gnc-fq-check`, `gnc-fq-dump`, `gnc-fq-helper`.

* `gnc-fq-update` stands a bit on its own. It's only purpose is to properly install Finance::Quote and its required perl modules.
* `gnc-fq-check` and `gnc-fq-helper` are used by the code to interface with Finance::Quote and are not intended to be called by the user directly.
* `gnc-fq-dump` is a command line tool to test how quotes are fetched and  dump the results. That only works properly if the ALPHAVANTAGE key has been set in the environment properly.

I'm inclined to change the code such that the user never has to call any of the last 3 scripts directly. The first two of those were not intended for users anyway, the later should be wrapped by gnucash-cli. The reason for this is that ensures the dump always happens in a consistent and controlled environment. Meaning, if the  dump via gnucash-cli works, fetching quotes in gnucash should equally work as they happen in the same environment (with proper environment variables set). If not, the dump will give failure information in the same context as where gnucash would fail. That should  help a bit with user support.

By the way gnucash-cli already exposes the outcome of gnucash-cli now, which is one less reason to call that script directly.

With this in mind I also plan to move the installation destination of these 3 scripts to <prefix>/share/gnucash/quotes rather than <prefix>/bin. For now gnc-fq-update could remain in bin as that's the only script I envision being run by users directly.

Circling back to your original suggestion, I agree with your proposal to make it part of app-utils. I see it more as convenience code rather than core business functionality. I think while at it, it may be useful to move the quotes directory with the perl scripts to become a subdirectory of app-utils as these scripts are now (mostly) considered internal implementation details.

A final consideration: Windows and macOS  users currently have to go through a few hoops to get Finance::Quote support on their systems. Linux users can just use their package manager instead. And flatpak has the necessary perl modules installed by default. Perhaps time has come to ship these by default in our installers/bundles on Windows and macOS as well ? I have no idea whether the security measures on macOS would allow this though. I also understand installing these modules involves compilation against the perl version to use. So that would probably mean packaging our own perl interpreter. I have no idea how much of a size increase that entails and whether that's allowed on macOS. So at this point it's an open question.

Reply:
> Circling back to your original suggestion, I agree with your proposal to make it part of app-utils. I see it more as convenience code rather than core business functionality. I think while at it, it may be useful to move the quotes directory with the perl scripts to become a subdirectory of app-utils as these scripts are now (mostly) considered internal implementation details.

On the other hand if we want gnc-commodity.c(pp) to be able to use a GncQuotes object, it has to be part of engine. engine can't depend on app-utils. So the proposal above changes to move the sources into engine and make the quotes directory with the perl helper scripts a subdirectory of engine.

Reply:
>     * That only works properly if the ALPHAVANTAGE key has been set in the environment properly.

1. Only if you are using that source.
2. There are changes on the way: https://lists.gnucash.org/pipermail/gnucash-devel/2020-November/045569.html.

Reply:
> /bin. For now gnc-fq-update could remain in bin as that's the only script I envision being run by users directly.

Formally it would belong to /usr/sbin.

Reply:
> If not, the dump will give failure information in the same context as where gnucash would fail.

`gnc-fq-dump -v …` gives much more information than GnuCash's `unknown system error`.

Reply:
> On the other hand if we want gnc-commodity.c(pp) to be able to use a GncQuotes object, it has to be part of engine.

I see it more as an import module like aqbanking.

Reply:
> > ```
> > * That only works properly if the ALPHAVANTAGE key has been set in the environment properly.
> > ```
> 
>     1. Only if you are using that source.
> 
>     2. There are changes on the way: https://lists.gnucash.org/pipermail/gnucash-devel/2020-November/045569.html.

I'm aware those changes are coming.  The idea remains the  same though: write a common backend for all the ways we support fetching quotes and have all frontends use that. That provides a more consistent user experience overall. AlphaVantage was just one example. If there will be other modules requiring other parameters to be set in other ways, all that then has to be encapsulated only once. Compare with the current situation: we have to tell our users that depend on AlphaVantage to set an environment  variable before running gnc-fq-dump and to set the API key preference in order for gnucash to be able to fetch quotes. That's not very helpful.

Reply:
> > /bin. For now gnc-fq-update could remain in bin as that's the only script I envision being run by users directly.
> 
> Formally it would belong to /usr/sbin.

In theory on linux yes, but the tool is not  intended for linux users. Storing it in prefix/sbin would make it a lonely tool in that directory...

Reply:
> > If not, the dump will give failure information in the same context as where gnucash would fail.
> 
> `gnc-fq-dump -v …` gives much more information than GnuCash's `unknown system error`.

That's part of what is about to change. By wrapping that file we can capture that same information in  gnucash and present it to the user directly. No more need to go to the command line.

Reply:
> > > If not, the dump will give failure information in the same context as where gnucash would fail.
> > 
> > 
> > `gnc-fq-dump -v …` gives much more information than GnuCash's `unknown system error`.
> 
> That's part of what is about to change. By wrapping that file we can capture that same information in gnucash and present it to the user directly. No more need to go to the command line.

But for clarity, there will still be a command line interface as well, only now part of gnucash-cli.

Reply:
> I see it more as an import module like aqbanking.

That's reasonable. I suppose it could go in libgnc-generic-import as easily as libgnc-app-utils. That seals not passing a `GncQuotes&` to `gnc_quote_source_set_fq_installed`, but on further reflection I think that it's better to avoid that coupling.

Reply:
> In theory on linux yes, but the tool is not intended for linux users.

Not so. Linux distros can't be relied up on to have current versions of F::Q in their package managers and often older ones are broken because the source sites have changed. There's a lot of web page scraping in F::Q.

Reply:
> A final consideration: Windows and macOS users currently have to go through a few hoops to get Finance::Quote support on their systems. Linux users can just use their package manager instead. And flatpak has the necessary perl modules installed by default. Perhaps time has come to ship these by default in our installers/bundles on Windows and macOS as well ? I have no idea whether the security measures on macOS would allow this though. I also understand installing these modules involves compilation against the perl version to use. So that would probably mean packaging our own perl interpreter. I have no idea how much of a size increase that entails and whether that's allowed on macOS. So at this point it's an open question.

I don't think anything links against libperl. Some of F::Q's dependencies have C components that link against system libraries, e.g. LWP links against the network stack. I've considered from time to time building those from their tarballs and including them in the mac bundle, but setting it up and maintaining it would be a fair amount of work because I wouldn't be able to use CPAN. We could link gnucash against libperl and not shell out... nah, let's not.

Reply:
> > A final consideration: Windows and macOS users currently have to go through a few hoops to get Finance::Quote support on their systems. Linux users can just use their package manager instead. And flatpak has the necessary perl modules installed by default. Perhaps time has come to ship these by default in our installers/bundles on Windows and macOS as well ? I have no idea whether the security measures on macOS would allow this though. I also understand installing these modules involves compilation against the perl version to use. So that would probably mean packaging our own perl interpreter. I have no idea how much of a size increase that entails and whether that's allowed on macOS. So at this point it's an open question.
> 
> I don't think anything links against libperl. Some of F::Q's dependencies have C components that link against system libraries, e.g. LWP links against the network stack. I've considered from time to time building those from their tarballs and including them in the mac bundle, but setting it up and maintaining it would be a fair amount of work because I wouldn't be able to use CPAN. We could link gnucash against libperl and not shell out... nah, let's not.

The modules needed for building Finance::Quote are listed in the flatpak modules. The build script for those (at least on  linux) is not that complicated  at first sight. Of course it would mean keeping track of all version updates.

Reply:
> > I see it more as an import module like aqbanking.
> 
> That's reasonable. I suppose it could go in libgnc-generic-import as easily as libgnc-app-utils. That seals not passing a `GncQuotes&` to `gnc_quote_source_set_fq_installed`, but on further reflection I think that it's better to avoid that coupling.

I have a few arguments to prefer putting it engine:
1. commodities have a quote source member variable. gnc-commodity internally maintains a list of quote sources and maps this against the list provided by Finance::Quote. So quote sources are deeply embedded in gnc-commodity.
2. I would like to expose this functionality in bindings (python or guile). While we (will) have some options for quote handling in gnucash-cli, that interface can never be as flexible as an api. If the quotes api is exposed via the bindings, this could be used in scenarios we are not thinking of ourselves. It's a gui free api so it lends itself nicely to be exposed via bindings.
3. I can see it as an import module though on a technical level there are differences: aqbanking support is a build time choice. Finance::Quote support on the other hand is a runtime choice. So the optionality of it is different. The code to handle Finance::Quote must be there at build time anyway, regardless of whether the user installed F::Q or not.

Having said all that, assume we treat it as an importer. I still want it exposed in bindings. So perhaps we should make an import-export directory in libgnucash to store import apis. That could the hold the Finance::Quote wrappers and also the api parts of the csv importers. Only the GUI components of these need to be in gnucash. In the future other importers could get rewritten in a similar way. Would that be an acceptable approach ?

Reply:
I see most of those arguments applying to all of the importers, of which only AQBanking is a build time option.

We're still a bit squishy on what exactly belongs in libgnucash, perhaps because we're still a bit squishy about what we want to eventually support as a library. As for gnucash containing only the GUI stuff, *you wish*. Even gnucash-cli.cpp is in there! I won't repeat my MVC rant, you've heard it enough to have memorized it. ;-) The price and commodity code is a bit too interlocked and could be made more testable with some separation. All of which is off the topic of this PR, which is about replacing price-quotes.scm. Let's keep the focus on that. At some point we should have a discussion about the overall organization and whether we really want to publish and maintain a separate libgnucash product, what should go into it, what the API should be, etc. and then get the conclusions into a wiki page.

In the meantime let's keep the focus of this PR on replacing price-quotes.scm. I don't think that it matters whether it goes in engine or app-utils, and as I said elsewhere I realize that passing a `GncQuotes&` to `gnc_quote_source_set_fq_availability` isn't a good idea.

Reply:
It has moved to app-utils for now as it needs functions from there  currently.